File: src/test/java/ru/jerael/booktracker/backend/api/validator/FileValidatorTest.java
package ru.jerael.booktracker.backend.api.validator;

import org.junit.jupiter.api.Test;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockMultipartFile;
import ru.jerael.booktracker.backend.domain.exception.ValidationException;
import ru.jerael.booktracker.backend.domain.exception.code.ErrorCode;
import static org.junit.jupiter.api.Assertions.*;

class FileValidatorTest {
    private final FileValidator fileValidator = new FileValidator();

    private final String fieldName = "cover";
    private final String fileName = "image.jpg";
    private final String contentType = MediaType.IMAGE_JPEG_VALUE;
    private final byte[] content = "content".getBytes();

    @Test
    void validate_ShouldPassValidationForValidFile() {
        MockMultipartFile mockMultipartFile = new MockMultipartFile("cover", fileName, contentType, content);

        assertDoesNotThrow(() -> fileValidator.validate(mockMultipartFile, fieldName));
    }

    @Test
    void validate_WhenFileIsNull_ShouldThrowValidationException() {
        ErrorCode errorCode =
            assertThrows(ValidationException.class, () -> fileValidator.validate(null, fieldName)).getErrorCode();

        assertEquals("EMPTY_FILE_CONTENT", errorCode.name());
    }

    @Test
    void validate_WhenFileIsEmpty_ShouldThrowValidationException() {
        MockMultipartFile mockMultipartFile = new MockMultipartFile("cover", fileName, contentType, new byte[0]);
        ErrorCode errorCode = assertThrows(ValidationException.class,
            () -> fileValidator.validate(mockMultipartFile, fieldName)).getErrorCode();

        assertEquals("EMPTY_FILE_CONTENT", errorCode.name());
    }

    @Test
    void validate_WhenFileNameIsNull_ShouldThrowValidationException() {
        MockMultipartFile mockMultipartFile = new MockMultipartFile("cover", null, contentType, content);
        ErrorCode errorCode = assertThrows(ValidationException.class,
            () -> fileValidator.validate(mockMultipartFile, fieldName)).getErrorCode();

        assertEquals("EMPTY_FILE_NAME", errorCode.name());
    }

    @Test
    void validate_WhenFileNameIsBlank_ShouldThrowValidationException() {
        MockMultipartFile mockMultipartFile = new MockMultipartFile("cover", "   ", contentType, content);
        ErrorCode errorCode = assertThrows(ValidationException.class,
            () -> fileValidator.validate(mockMultipartFile, fieldName)).getErrorCode();

        assertEquals("EMPTY_FILE_NAME", errorCode.name());
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/validator/FileValidatorTest.kt
package ru.jerael.booktracker.backend.web.validator

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import ru.jerael.booktracker.backend.domain.exception.ValidationException
import ru.jerael.booktracker.backend.factory.file.FileFactory

class FileValidatorTest {
    private val validator = FileValidator()
    
    @Test
    fun `validate should pass validation for valid file`() {
        val file = FileFactory.createMockMultipartFile()
        
        assertDoesNotThrow { validator.validate(file, "cover") }
    }
    
    @Test
    fun `when file is null, validate should throw ValidationException`() {
        val file = null
        
        val exception = assertThrows(ValidationException::class.java) {
            validator.validate(file, "cover")
        }
        
        assertThat(exception.errors).anyMatch { it.field.equals("cover") }
    }
    
    @Test
    fun `when file is empty, validate should throw ValidationException`() {
        val file = FileFactory.createMockMultipartFile(content = ByteArray(0))
        
        val exception = assertThrows(ValidationException::class.java) {
            validator.validate(file, "cover")
        }
        
        assertThat(exception.errors).anyMatch { it.field.equals("cover") }
    }
    
    @Test
    fun `when fileName is empty, validate should throw ValidationException`() {
        val file = FileFactory.createMockMultipartFile(originalFileName = "")
        
        val exception = assertThrows(ValidationException::class.java) {
            validator.validate(file, "cover")
        }
        
        assertThat(exception.errors).anyMatch { it.field.equals("cover") }
    }
    
    @Test
    fun `when fileName is blank, validate should throw ValidationException`() {
        val file = FileFactory.createMockMultipartFile(originalFileName = "  ")
        
        val exception = assertThrows(ValidationException::class.java) {
            validator.validate(file, "cover")
        }
        
        assertThat(exception.errors).anyMatch { it.field.equals("cover") }
    }
}